File: concurrency-async/concurrency_async.py
import asyncio
import threading
import time
from concurrent.futures import ThreadPoolExecutor


# --- Threads ---

def worker(name: str, delay: float) -> str:
    time.sleep(delay)
    return f"Thread '{name}' done after {delay}s"


def run_threads() -> None:
    print("\n-- Threads --")
    with ThreadPoolExecutor(max_workers=3) as executor:
        futures = [
            executor.submit(worker, "A", 0.2),
            executor.submit(worker, "B", 0.1),
            executor.submit(worker, "C", 0.05),
        ]
        for f in futures:
            print(f.result())


# --- async/await (event loop model) ---

async def fetch_data(name: str, delay: float) -> str:
    await asyncio.sleep(delay)   # non-blocking wait
    return f"Data from '{name}'"


async def run_async() -> None:
    print("\n-- async/await (event loop) --")

    # Sequential
    result_a = await fetch_data("source-A", 0.1)
    result_b = await fetch_data("source-B", 0.1)
    print("Sequential:", result_a, "|", result_b)

    # Parallel with asyncio.gather
    results = await asyncio.gather(
        fetch_data("source-X", 0.2),
        fetch_data("source-Y", 0.1),
        fetch_data("source-Z", 0.05),
    )
    print("Parallel gather:", results)


def main() -> None:
    print("Language: Python")

    # Threads (OS threads, but GIL limits CPU parallelism for pure Python)
    run_threads()

    # Event loop — single-threaded cooperative concurrency
    asyncio.run(run_async())


if __name__ == "__main__":
    main()

File: concurrency-async/concurrencyAsync.js
// JavaScript runs on a single-threaded event loop.
// True parallelism requires Worker Threads (Node.js) or Web Workers (browser).

const { Worker, isMainThread, parentPort, workerData } = require("worker_threads");

// --- async/await & Promises (event loop model) ---

async function fetchData(name, delayMs) {
  await new Promise((resolve) => setTimeout(resolve, delayMs));
  return `Data from '${name}'`;
}

async function runAsync() {
  console.log("\n-- async/await (event loop) --");

  // Sequential: each await waits before the next starts
  const a = await fetchData("source-A", 100);
  const b = await fetchData("source-B", 100);
  console.log("Sequential:", a, "|", b);

  // Parallel: Promise.all runs concurrently on the event loop
  const results = await Promise.all([
    fetchData("source-X", 200),
    fetchData("source-Y", 100),
    fetchData("source-Z", 50),
  ]);
  console.log("Promise.all parallel:", results);

  // Promise.allSettled — collects results even if some reject
  const settled = await Promise.allSettled([
    fetchData("ok", 50),
    Promise.reject(new Error("simulated failure")),
  ]);
  settled.forEach((r) =>
    console.log("allSettled:", r.status, r.value ?? r.reason?.message)
  );
}

// --- Worker Threads (true parallelism in Node.js) ---

function runWorkerThread() {
  return new Promise((resolve, reject) => {
    const worker = new Worker(
      `
      const { parentPort, workerData } = require('worker_threads');
      // CPU-bound work isolated from the main thread
      let sum = 0;
      for (let i = 0; i < workerData.count; i++) sum += i;
      parentPort.postMessage({ thread: workerData.name, sum });
      `,
      { eval: true, workerData: { name: "worker-1", count: 1_000_000 } }
    );
    worker.on("message", resolve);
    worker.on("error", reject);
  });
}

async function main() {
  console.log("Language: JavaScript");

  await runAsync();

  console.log("\n-- Worker Thread (parallel CPU work) --");
  const result = await runWorkerThread();
  console.log(`Worker result: thread=${result.thread}, sum=${result.sum}`);
}

main();
